<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.43">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Ryan M. Moore, PhD">
<meta name="dcterms.date" content="2025-04-09">
<meta name="description" content="Learn essential Python programming skills tailored for life science research. This practical guide helps biology and life science students with no prior coding experience develop fundamental programming and data analysis skills.">

<title>8&nbsp; Intro to Statistics &amp; Modeling with Python – Practical Python Programming for Life Scientists</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../chapters/09_io_files_contexts.html" rel="next">
<link href="../chapters/07_exploratory_data_analysis.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-d4d76bf8491c20bad77d141916dc28e1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-7b496870f2c3c580a869e15250cb8aa1.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>
<meta name="google-site-verification" content="lAMnqKIoNocjUOHIuFkaXcMzcceqyGBcevAs7t3N_AI">

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<meta property="og:title" content="8&nbsp; Intro to Statistics &amp; Modeling with Python – Practical Python Programming for Life Scientists">
<meta property="og:description" content="Learn essential Python programming skills tailored for life science research. This practical guide helps biology and life science students with no prior coding experience develop fundamental programming and data analysis skills.">
<meta property="og:image" content="https://mooreryan.github.io/practical_python/chapters/08_stats_models_files/figure-html/cell-3-output-1.png">
<meta property="og:site_name" content="Practical Python Programming for Life Scientists">
<meta property="og:image:height" content="278">
<meta property="og:image:width" content="451">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../chapters/08_stats_models.html"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Intro to Statistics &amp; Modeling with Python</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header sidebar-header-stacked">
      <a href="../index.html" class="sidebar-logo-link">
      <img src="../images/dna_python_logo.webp" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Practical Python Programming for Life Scientists</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Introduction</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/01_basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Basics</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/02_collections.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Collections</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/03_algorithms.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Algorithmic Thinking</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/04_functions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Functions</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/05_classes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Introduction to Object-Oriented Programming</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/06_errors.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Errors &amp; Exceptions</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/07_exploratory_data_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Intro to Exploratory Data Analysis with Python</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/08_stats_models.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Intro to Statistics &amp; Modeling with Python</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/09_io_files_contexts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">I/O, Files, &amp; Contexts</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A01_solutions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendix 1: Solutions</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A02_regex.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Appendix 2: Regular Expressions</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#setup" id="toc-setup" class="nav-link" data-scroll-target="#setup">Setup</a></li>
  <li><a href="#statistical-tests" id="toc-statistical-tests" class="nav-link" data-scroll-target="#statistical-tests">Statistical Tests</a>
  <ul>
  <li><a href="#hypothesis-testing-and-p-values" id="toc-hypothesis-testing-and-p-values" class="nav-link" data-scroll-target="#hypothesis-testing-and-p-values">Hypothesis Testing and p-values</a></li>
  <li><a href="#brief-glossary" id="toc-brief-glossary" class="nav-link" data-scroll-target="#brief-glossary">Brief Glossary</a></li>
  <li><a href="#comparing-two-groups" id="toc-comparing-two-groups" class="nav-link" data-scroll-target="#comparing-two-groups">Comparing Two Groups</a>
  <ul>
  <li><a href="#unpaired-groups" id="toc-unpaired-groups" class="nav-link" data-scroll-target="#unpaired-groups">Unpaired Groups</a>
  <ul>
  <li><a href="#unpaired-t-test" id="toc-unpaired-t-test" class="nav-link" data-scroll-target="#unpaired-t-test">Unpaired t Test</a></li>
  <li><a href="#mann-whitney-test" id="toc-mann-whitney-test" class="nav-link" data-scroll-target="#mann-whitney-test">Mann-Whitney Test</a></li>
  </ul></li>
  <li><a href="#paired-groups" id="toc-paired-groups" class="nav-link" data-scroll-target="#paired-groups">Paired Groups</a>
  <ul>
  <li><a href="#paired-t-test" id="toc-paired-t-test" class="nav-link" data-scroll-target="#paired-t-test">Paired t test</a>
  <ul class="collapse">
  <li><a href="#aside-effect-size-and-p-values" id="toc-aside-effect-size-and-p-values" class="nav-link" data-scroll-target="#aside-effect-size-and-p-values">Aside: Effect Size and p-values</a></li>
  </ul></li>
  <li><a href="#wilcoxon-test" id="toc-wilcoxon-test" class="nav-link" data-scroll-target="#wilcoxon-test">Wilcoxon test</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#comparing-multiple-groups" id="toc-comparing-multiple-groups" class="nav-link" data-scroll-target="#comparing-multiple-groups">Comparing Multiple Groups</a>
  <ul>
  <li><a href="#one-way-anova" id="toc-one-way-anova" class="nav-link" data-scroll-target="#one-way-anova">One-Way ANOVA</a></li>
  <li><a href="#kruskal-wallis" id="toc-kruskal-wallis" class="nav-link" data-scroll-target="#kruskal-wallis">Kruskal-Wallis</a></li>
  <li><a href="#post-hoc-testing" id="toc-post-hoc-testing" class="nav-link" data-scroll-target="#post-hoc-testing">Post-Hoc Testing</a>
  <ul>
  <li><a href="#tukeys-hsd-with-scipy" id="toc-tukeys-hsd-with-scipy" class="nav-link" data-scroll-target="#tukeys-hsd-with-scipy">Tukey’s HSD with SciPy</a></li>
  <li><a href="#tukeys-hsd-with-statsmodels" id="toc-tukeys-hsd-with-statsmodels" class="nav-link" data-scroll-target="#tukeys-hsd-with-statsmodels">Tukey’s HSD with statsmodels</a></li>
  </ul></li>
  </ul></li>
  </ul></li>
  <li><a href="#linear-regression" id="toc-linear-regression" class="nav-link" data-scroll-target="#linear-regression">Linear Regression</a>
  <ul>
  <li><a href="#with-no-relationship-in-the-data" id="toc-with-no-relationship-in-the-data" class="nav-link" data-scroll-target="#with-no-relationship-in-the-data">With No Relationship in the Data</a></li>
  <li><a href="#multiple-predictors" id="toc-multiple-predictors" class="nav-link" data-scroll-target="#multiple-predictors">Multiple Predictors</a></li>
  </ul></li>
  <li><a href="#dimensionality-reduction-clustering" id="toc-dimensionality-reduction-clustering" class="nav-link" data-scroll-target="#dimensionality-reduction-clustering">Dimensionality Reduction &amp; Clustering</a>
  <ul>
  <li><a href="#principal-components-analysis-pca" id="toc-principal-components-analysis-pca" class="nav-link" data-scroll-target="#principal-components-analysis-pca">Principal Components Analysis (PCA)</a></li>
  <li><a href="#k-means-clustering" id="toc-k-means-clustering" class="nav-link" data-scroll-target="#k-means-clustering">k-means Clustering</a></li>
  </ul></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary">Summary</a></li>
  <li><a href="#practice-problems" id="toc-practice-problems" class="nav-link" data-scroll-target="#practice-problems">Practice Problems</a>
  <ul>
  <li><a href="#sec-problem-8.1" id="toc-sec-problem-8.1" class="nav-link" data-scroll-target="#sec-problem-8.1">8.1</a></li>
  <li><a href="#sec-problem-8.2" id="toc-sec-problem-8.2" class="nav-link" data-scroll-target="#sec-problem-8.2">8.2</a></li>
  <li><a href="#sec-problem-8.3" id="toc-sec-problem-8.3" class="nav-link" data-scroll-target="#sec-problem-8.3">8.3</a></li>
  <li><a href="#sec-problem-8.4" id="toc-sec-problem-8.4" class="nav-link" data-scroll-target="#sec-problem-8.4">8.4</a></li>
  <li><a href="#sec-problem-8.5" id="toc-sec-problem-8.5" class="nav-link" data-scroll-target="#sec-problem-8.5">8.5</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Intro to Statistics &amp; Modeling with Python</span></h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Ryan M. Moore, PhD </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">April 9, 2025</p>
    </div>
  </div>
  
    <div>
    <div class="quarto-title-meta-heading">Modified</div>
    <div class="quarto-title-meta-contents">
      <p class="date-modified">April 18, 2025</p>
    </div>
  </div>
    
  </div>
  


</header>


<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>Python has become a cornerstone tool for statistical analysis in life sciences, offering powerful yet accessible ways to understand biological data. Statistical analysis helps scientists move beyond simple observations to make reliable inferences about experimental results and natural phenomena. Python’s data science libraries (like NumPy, Pandas, and SciPy) provide ready-to-use functions that handle complex statistical calculations without requiring advanced mathematical knowledge. Modeling biological processes mathematically allows scientists to test hypotheses, predict outcomes, and discover patterns that might not be visible through direct observation. For life scientists, building statistical literacy is as important as lab techniques–it’s how we determine if results are meaningful or merely coincidental. Python’s visualization capabilities (through libraries like Matplotlib and Seaborn) help make abstract statistical concepts more concrete and interpretable. Understanding basic statistics in Python empowers biologists to design better experiments, analyze results more critically, and communicate findings more effectively.</p>
</section>
<section id="setup" class="level2">
<h2 class="anchored" data-anchor-id="setup">Setup</h2>
<p>Ensure that you have the following packages installed:</p>
<ul>
<li><a href="https://matplotlib.org/">Matplotlib</a></li>
<li><a href="https://numpy.org/">NumPy</a></li>
<li><a href="https://pandas.pydata.org/">Pandas</a></li>
<li><a href="https://scipy.org/">SciPy</a></li>
<li><a href="https://seaborn.pydata.org/">Seaborn</a></li>
<li><a href="https://www.statsmodels.org/dev/index.html">statsmodels</a></li>
</ul>
<p>Next, import all the required items for the chapter. We will use the common abbreviations when appropriate.</p>
<div id="05dfe391" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> statsmodels.formula.api</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> statsmodels.stats.multicomp</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collections <span class="im">import</span> Counter</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy <span class="im">import</span> cluster</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy <span class="im">import</span> stats</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> statsmodels.multivariate.pca <span class="im">import</span> PCA</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>pd.set_option(<span class="st">"display.max_rows"</span>, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="statistical-tests" class="level2">
<h2 class="anchored" data-anchor-id="statistical-tests">Statistical Tests</h2>
<p>Statistical tests are mathematical methods that help researchers determine whether patterns in their data represent genuine relationships or simply random chance.</p>
<p><em>Note: Statistical tests all have caveats and other considerations that you need to be aware of before running the test and while interpreting the output. I suggest that you carefully read the docs for each of these functions as well as consult any relevant statistical literature when designing and analyzing your experiments.</em></p>
<section id="hypothesis-testing-and-p-values" class="level3">
<h3 class="anchored" data-anchor-id="hypothesis-testing-and-p-values">Hypothesis Testing and p-values</h3>
<p>The goal of a statistical test is to determine whether a specific hypothesis about data is likely true. When conducting such a test, we set up two hypotheses: the null hypothesis (H0) and the alternative hypothesis (HA). The null hypothesis typically states that there’s no effect or relationship in our data (essentially, “nothing interesting is happening”), while the alternative hypothesis suggests that some effect or relationship does exist. If our test results show statistical significance, we reject the null hypothesis in favor of the alternative. However, if we don’t find significance, we fail to reject the null hypothesis (rather than “accepting” it, a subtle but important distinction). To determine significance, we calculate a p-value, which tells us the probability of observing our test results (or something more extreme) if the null hypothesis were actually true. Before conducting the test, we choose a significance level (commonly 0.05) as our threshold. This means we’re comfortable with a 5% risk of rejecting a null hypothesis that is actually true (known as a Type I error or false positive). On the flip side, we might also fail to reject a null hypothesis that’s actually false. This is called a Type II error, and it happens when our test misses a real effect that exists in the data (false negative).</p>
<p>For more on hypothesis testing and p-values, see <a href="https://online.stat.psu.edu/stat500/lesson/6a/6a.1">Introduction to Hypothesis Testing</a> and <a href="https://online.stat.psu.edu/stat500/lesson/6a/6a.2">Steps for Hypothesis Tests</a>.</p>
<p>This is probably a good time to mention that p-values have been controversial for quite a while now, for examples, see:</p>
<ul>
<li><a href="https://doi.org/10.1080/00031305.2019.1583913">Moving to a World Beyond “<em>p</em> &lt; 0.05”</a></li>
<li><a href="https://doi.org/10.1007/s10654-010-9440-x">The ongoing tyranny of statistical significance testing in biomedical research</a></li>
<li><a href="https://doi.org/10.1053/j.seminhematol.2008.04.003">A Dirty Dozen: Twelve P-Value Misconceptions</a></li>
</ul>
<p>There are <em>many</em> more examples to be found in the literature about this topic!</p>
<p><em>Note: For this document, we will use a p-value cutoff of 0.05, and talk about significance and rejecting null hypotheses in as basic a way as possible, but you should keep the above criticism in mind.</em></p>
<div id="tip-08-pvalue-misconceptions" class="callout callout-style-default callout-tip callout-titled" title="Stop &amp; Think">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip&nbsp;8.1: Stop &amp; Think
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>What’s one potential problem with relying solely on p-values when interpreting experimental results?</p>
</div>
</div>
</div>
</section>
<section id="brief-glossary" class="level3">
<h3 class="anchored" data-anchor-id="brief-glossary">Brief Glossary</h3>
<p>We’re going to touch briefly on some common statistical tests. Before we do, let’s review a few terms:</p>
<ul>
<li><strong>Normal distribution:</strong> Data is <a href="https://www.khanacademy.org/math/statistics-probability/modeling-distributions-of-data/normal-distributions-library/a/normal-distributions-review">normally distributed</a> if it fits a symmetrical bell curve and the mean and median are both at the center of the distribution.</li>
<li><strong>Independent sample:</strong> Sample groups are independent if observations do not influence each other.</li>
<li><strong>Paired sample:</strong> Samples are paired (or dependent) when there is a meaningful relationship between observations in two or more groups. These can be measurements of the same individual over time or related measurements of two or more individuals. There are good examples of when this might happen <a href="https://online.stat.psu.edu/stat200/lesson/1/1.4/1.4.3">here</a> and <a href="https://www.statisticssolutions.com/free-resources/directory-of-statistical-analyses/paired-sample-t-test/">here</a>.</li>
<li><strong>Assumption:</strong> All statistical tests assume things about your data in their calculations. The big ones are usual the <a href="https://real-statistics.com/descriptive-statistics/assumptions-statistical-test/">distribution, independence, skewness, and linearity</a>. Performing statistical tests on data that does not meet the correct assumptions can lead to erroneous results, though this may vary with sample size. Statistical tests generally fall into two categories (see the articles <a href="https://www.mayo.edu/research/documents/parametric-and-nonparametric-demystifying-the-terms/doc-20408960">Parametric and Nonparametric: Demystifying the Terms</a> and <a href="https://doi.org/10.36416/1806-3756/e20210292">Nonparametric statistical tests: friend or foe?</a> for more thorough explanations and comparisons):
<ul>
<li><strong>Parametric tests:</strong> These assume specific data characteristics, generally a normal distribution. These assumptions matter most with smaller sample sizes.</li>
<li><strong>Nonparametric tests:</strong> These make fewer assumptions about your data’s shape and other parameters. They often work by analyzing ranks rather than raw values.</li>
</ul></li>
</ul>
</section>
<section id="comparing-two-groups" class="level3">
<h3 class="anchored" data-anchor-id="comparing-two-groups">Comparing Two Groups</h3>
<p>The statistical comparison of two groups is one of the more common operations that you will need to perform. This covers everything from comparing gene expression levels between healthy and diseased tissue to measuring enzyme activity before and after drug treatment. Comparing two groups is mostly done with the venerable t test, as well as the Mann-Whitney and Wilcoxon tests when the underlying data is not sampled from a normal distribution.</p>
<p>One distinction that needs to be made is in the experimental design, that is, whether your groups are paired or not. Paired groups are measurements taken from the same subjects under different conditions (like before and after treatment). Alternatively, unpaired groups are measurements from different subjects (like control group versus experimental group).</p>
<p>We will start with unpaired groups, since dealing with them is a bit simpler.</p>
<section id="unpaired-groups" class="level4">
<h4 class="anchored" data-anchor-id="unpaired-groups">Unpaired Groups</h4>
<p>Comparing two unpaired groups is a classic problem, and you have probably needed this operation many times in your own research. Let’s start by looking at the unpaired <a href="https://www.jmp.com/en/statistics-knowledge-portal/t-test">t test</a>.</p>
<section id="unpaired-t-test" class="level5">
<h5 class="anchored" data-anchor-id="unpaired-t-test">Unpaired t Test</h5>
<p>An unpaired t test, also known as a <a href="https://www.jmp.com/en/statistics-knowledge-portal/t-test/two-sample-t-test">two-sample t test</a>, compares means between two independent groups. This is the classic t test that you may remember from your statistics courses. Let’s create some example data and visualize it to see how it works.</p>
<div id="2d25f71a" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Ensures the same random numbers are generated each time.</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>np.random.seed(<span class="dv">9382741</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a DataFrame with two columns of random data:</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame(</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>        <span class="co"># "Group A": Generate 40 random values from a normal distribution</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>        <span class="co"># with mean (loc) of 5 and standard deviation (scale) of 3</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Group A"</span>: stats.norm(loc<span class="op">=</span><span class="dv">5</span>, scale<span class="op">=</span><span class="dv">3</span>).rvs(size<span class="op">=</span><span class="dv">40</span>),</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>        <span class="co"># "Group B": Generate 40 random values from a normal distribution</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>        <span class="co"># with mean (loc) of 8 and standard deviation (scale) of 3</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Group B"</span>: stats.norm(loc<span class="op">=</span><span class="dv">8</span>, scale<span class="op">=</span><span class="dv">3</span>).rvs(size<span class="op">=</span><span class="dv">40</span>),</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a categorical strip plot (jittered points) using seaborn</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="co"># - 'kind="strip"' specifies we want individual data points with jitter</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a><span class="co"># - 'height=3' sets the height of the plot in inches</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="co"># - 'aspect=1.6' sets the width-to-height ratio</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>sns.catplot(df, kind<span class="op">=</span><span class="st">"strip"</span>, height<span class="op">=</span><span class="dv">3</span>, aspect<span class="op">=</span><span class="fl">1.6</span>)</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Overlay a boxplot on the same axes</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a><span class="co"># - 'width=0.25' makes the boxes narrower than default</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a><span class="co"># - 'fill=False' creates unfilled boxes (just outlines)</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a><span class="co"># - 'color="#555555"' sets the box color to a dark gray</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>sns.boxplot(df, width<span class="op">=</span><span class="fl">0.25</span>, fill<span class="op">=</span><span class="va">False</span>, color<span class="op">=</span><span class="st">"#555555"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="08_stats_models_files/figure-html/cell-3-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Now that we have seen the data, let’s run the t test. To do it, we use the <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html#scipy.stats.ttest_ind">ttest_ind()</a> function from <a href="https://docs.scipy.org/doc/scipy/index.html">SciPy</a>.</p>
<div id="5e2cfc9e" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform an unpaired t test comparing Group A and Group B</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># stats.ttest_ind() performs the independent samples t test</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># df["Group A"] and df["Group B"] are the two samples we're comparing</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># .pvalue extracts just the p-value from the test results</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> stats.ttest_ind(df[<span class="st">"Group A"</span>], df[<span class="st">"Group B"</span>]).pvalue</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the mean value for each column in the data frame</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co"># - df.mean() calculates the mean of each column in the original data frame</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="co"># - columns=["Mean"] sets the name of the resulting column to "Mean"</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>means <span class="op">=</span> pd.DataFrame(df.mean(), columns<span class="op">=</span>[<span class="st">"Mean"</span>])</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>display(means)</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Unpaired t test p-value: </span><span class="sc">{</span>result<span class="sc">:.1e}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Mean</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Group A</td>
<td>5.345036</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Group B</td>
<td>7.455855</td>
</tr>
</tbody>
</table>

</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Unpaired t test p-value: 1.5e-03</code></pre>
</div>
</div>
<p>The p-value of the t test is 1.5e-03 so we reject the null hypothesis that the group means are the same.</p>
<p>The t test comes with several important assumptions that should be met for the results to be valid:</p>
<ul>
<li>Samples are independent</li>
<li>Samples have equal variance (homoscedasticity or equal spread of data)</li>
<li>Data follows a normal distribution</li>
<li>Random sampling from the population</li>
</ul>
</section>
<section id="mann-whitney-test" class="level5">
<h5 class="anchored" data-anchor-id="mann-whitney-test">Mann-Whitney Test</h5>
<p>If your data doesn’t follow a normal distribution, you may need a nonparametric alternative like the <a href="https://www.statisticssolutions.com/free-resources/directory-of-statistical-analyses/mann-whitney-u-test/">Mann-Whitney Test</a>. Instead of comparing means, this test determines if values from one group tend to be larger than values from another group.</p>
<p>First, we generate and visualize some data. We will generate some data from a uniform distribution (non-normal), and each of the samples will be small (10 observations each). This is a pretty good use-case for the Mann-Whitney test.</p>
<div id="tip-08-parametric-vs-nonparametric" class="callout callout-style-default callout-tip callout-titled" title="Stop &amp; Think">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip&nbsp;8.2: Stop &amp; Think
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>When might you choose a nonparametric test like Mann-Whitney over a parametric test like the t test?</p>
</div>
</div>
</div>
<div id="75417145" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>np.random.seed(<span class="dv">9382741</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame(</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Group A"</span>: stats.uniform(loc<span class="op">=</span><span class="dv">5</span>, scale<span class="op">=</span><span class="dv">10</span>).rvs(size<span class="op">=</span><span class="dv">6</span>),</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Group B"</span>: stats.uniform(loc<span class="op">=</span><span class="dv">8</span>, scale<span class="op">=</span><span class="dv">10</span>).rvs(size<span class="op">=</span><span class="dv">6</span>),</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>sns.catplot(df, kind<span class="op">=</span><span class="st">"strip"</span>, height<span class="op">=</span><span class="dv">3</span>, aspect<span class="op">=</span><span class="fl">1.6</span>)</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>sns.boxplot(df, width<span class="op">=</span><span class="fl">0.25</span>, fill<span class="op">=</span><span class="va">False</span>, color<span class="op">=</span><span class="st">"#555555"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="08_stats_models_files/figure-html/cell-5-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Just from looking at the data, it appears that there might be a real difference between the two groups. Let’s run the Mann-Whitney test to check if they are significantly different.</p>
<div id="fa5154a2" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform the Mann-Whitney U test comparing Groups A and B</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (We extract just the p-value from the test results with the `.pvalue` at the end)</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> stats.mannwhitneyu(df[<span class="st">"Group A"</span>], df[<span class="st">"Group B"</span>]).pvalue</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the median value for each column in the data frame</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>medians <span class="op">=</span> pd.DataFrame(df.median(), columns<span class="op">=</span>[<span class="st">"Median"</span>])</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>display(medians)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Mann-Whitney test p-value: </span><span class="sc">{</span>result<span class="sc">:.2f}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Median</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Group A</td>
<td>8.414715</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Group B</td>
<td>12.917969</td>
</tr>
</tbody>
</table>

</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Mann-Whitney test p-value: 0.03</code></pre>
</div>
</div>
<p>Cool! So we see that the Mann-Whitney test returned a p-value &lt; 0.05 so we reject the null hypothesis that the distribution underlying group A is the same as the distribution underlying group B.</p>
<p>While the Mann-Whitney test doesn’t assume that the samples come from a normal distribution, it does still have assumptions that should be met and that can affect your interpretation of the results. For more info see <a href="https://www.statisticssolutions.com/free-resources/directory-of-statistical-analyses/mann-whitney-u-test/">here</a>.</p>
</section>
</section>
<section id="paired-groups" class="level4">
<h4 class="anchored" data-anchor-id="paired-groups">Paired Groups</h4>
<ul>
<li>For comparing the means of paired samples, you will want to use a <a href="https://www.jmp.com/en/statistics-knowledge-portal/t-test/paired-t-test">paired t test</a>.</li>
<li>The nonparametric companion to the paired t test is the <a href="https://datatab.net/tutorial/wilcoxon-test">Wilcoxon test</a>.</li>
</ul>
<section id="paired-t-test" class="level5">
<h5 class="anchored" data-anchor-id="paired-t-test">Paired t test</h5>
<p>We use the <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_rel.html#scipy.stats.ttest_rel">scipy.stats.ttest_rel()</a> function from SciPy to perform a paired t-test. This statistical test examines whether two related samples (like before-and-after measurements from the same subjects) have the same average values. The test evaluates the null hypothesis that the means of the two paired groups are identical.</p>
<p>Same as before, let’s create some data and make some box plots to check it out. In this dataset, we would expect to see about a 2 unit increase in the post treatment measurements.</p>
<div id="3e3caec9" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the random seed for reproducibility</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>np.random.seed(<span class="dv">964832</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate random pre-treatment measurements from a normal distribution</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co"># with mean=50 and standard deviation=10 for 40 subjects</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>before_treatment <span class="op">=</span> stats.norm(loc<span class="op">=</span><span class="dv">50</span>, scale<span class="op">=</span><span class="dv">10</span>).rvs(size<span class="op">=</span><span class="dv">40</span>)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a DataFrame with two columns:</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="co"># - 'Pre': the pre-treatment measurements</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="co"># - 'Post': the pre-treatment values plus a small random effect</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="co">#   (representing treatment effect) from a normal distribution</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a><span class="co">#   with mean=2 and standard deviation=1</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame(</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Pre"</span>: before_treatment,</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Post"</span>: before_treatment <span class="op">+</span> stats.norm(loc<span class="op">=</span><span class="dv">2</span>, scale<span class="op">=</span><span class="dv">1</span>).rvs(size<span class="op">=</span><span class="dv">40</span>),</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a strip plot (jittered points) to visualize the individual data points</span></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a><span class="co"># in both Pre and Post groups</span></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>sns.catplot(df, kind<span class="op">=</span><span class="st">"strip"</span>, height<span class="op">=</span><span class="dv">3</span>, aspect<span class="op">=</span><span class="fl">1.6</span>)</span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Add box plots over the strip plots to show summary statistics</span></span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a><span class="co"># width=0.25 makes the boxes narrow</span></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a><span class="co"># fill=False means the boxes are transparent</span></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a><span class="co"># color="#555555" sets the box color to a dark gray</span></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>sns.boxplot(df, width<span class="op">=</span><span class="fl">0.25</span>, fill<span class="op">=</span><span class="va">False</span>, color<span class="op">=</span><span class="st">"#555555"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="08_stats_models_files/figure-html/cell-7-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Next, we run the paired t test.</p>
<div id="843eb867" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run the t test</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> stats.ttest_rel(df[<span class="st">"Pre"</span>], df[<span class="st">"Post"</span>]).pvalue</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># This is to make a nice table with the means.</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>means <span class="op">=</span> pd.DataFrame(df.mean(), columns<span class="op">=</span>[<span class="st">"Mean"</span>])</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>display(means)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Unpaired t test p-value: </span><span class="sc">{</span>result<span class="sc">:.1e}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Mean</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Pre</td>
<td>50.150032</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Post</td>
<td>52.008654</td>
</tr>
</tbody>
</table>

</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Unpaired t test p-value: 3.3e-12</code></pre>
</div>
</div>
<p>It is statistically significant! But wait, while our p-value confirms this, we need to consider whether 2 units (whatever those units represent in your experiment) actually matters in your biological system. This is where your expertise as a researcher becomes crucial. It’s important not to rely solely on p-values when interpreting results. In our example, we had sufficient sample size to detect a difference, but that difference might still be too small to have any meaningful biological impact.</p>
<div id="tip-08-statistical-vs-biological" class="callout callout-style-default callout-tip callout-titled" title="Stop &amp; Think">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip&nbsp;8.3: Stop &amp; Think
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>Why is it important to consider both statistical significance and biological relevance when interpreting results?</p>
</div>
</div>
</div>
<p>As you can see, this analysis is quite similar to the unpaired test we discussed earlier. The main difference to keep in mind is how you interpret the results: with paired tests, you’re looking at changes within the same subjects (like before and after treatment), rather than differences between separate groups.</p>
<section id="aside-effect-size-and-p-values" class="level6">
<h6 class="anchored" data-anchor-id="aside-effect-size-and-p-values">Aside: Effect Size and p-values</h6>
<p>Scientists have a tendency to focus strongly on p-values, but there is another measure that arguably deserves more attention: <em>effect size</em>. Effect size tells us something different than a p-value: how large or meaningful that effect actually is. It measures the magnitude of differences between groups or the strength of relationships between variables in practical terms. In studies with large sample sizes, it’s very common to find statistically significant results (tiny p-values) alongside tiny effect sizes. This creates a situation where a difference might be technically “real” but practically meaningless. This is precisely why researchers often report both p-values and effect sizes together. P-values answer “Is there an effect?”, while effect sizes answer the equally crucial question: “Does this effect actually matter?” Together, they provide a much more complete and actionable understanding of research findings.</p>
<p>For more on effect sizes, see <a href="https://doi.org/10.4300/JGME-D-12-00156.1">Using Effect Size–or Why the <em>P</em> Value Is Not Enough</a>.</p>
<div id="tip-08-effect-size-examples" class="callout callout-style-default callout-tip callout-titled" title="Stop &amp; Think">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip&nbsp;8.4: Stop &amp; Think
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>Can you think of a scenario in the life sciences where a statistically significant result might have a negligible effect size?</p>
</div>
</div>
</div>
</section>
</section>
<section id="wilcoxon-test" class="level5">
<h5 class="anchored" data-anchor-id="wilcoxon-test">Wilcoxon test</h5>
<p>The Wilcoxon signed-rank test is used to test the null hypothesis that two related paired samples come from the same distribution. Here is what the SciPy docs have to say about the <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.wilcoxon.html#scipy.stats.wilcoxon">Wilcoxon signed-rank test</a>:</p>
<blockquote class="blockquote">
<p>The Wilcoxon signed-rank test tests the null hypothesis that two related paired samples come from the same distribution. In particular, it tests whether the distribution of the differences x - y is symmetric about zero. It is a non-parametric version of the paired t test.</p>
</blockquote>
<p><em>Note: this is pretty similar to the Mann-Whitney test…we will leave the distinction to your stats professors :)</em></p>
<p>Again, let’s make some fake data and see how it looks.</p>
<div id="fa9008ab" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set a random seed for reproducibility</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>np.random.seed(<span class="dv">9382741</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a DataFrame with two columns:</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. "Pre-Treatment": 6 random values from a uniform distribution</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co">#    between 5 and 11 (loc=5, scale=6)</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. "Post-Treatment": 6 random values from a uniform distribution</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="co">#    between 6 and 12 (loc=6, scale=6)</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame(</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>    {</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Pre-Treatment"</span>: stats.uniform(loc<span class="op">=</span><span class="dv">5</span>, scale<span class="op">=</span><span class="dv">6</span>).rvs(size<span class="op">=</span><span class="dv">6</span>),</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Post-Treatment"</span>: stats.uniform(loc<span class="op">=</span><span class="dv">6</span>, scale<span class="op">=</span><span class="dv">6</span>).rvs(size<span class="op">=</span><span class="dv">6</span>),</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a strip plot (individual data points) using Seaborn's catplot function</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a><span class="co"># - height=3 and aspect=1.6 control the dimensions of the plot</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>sns.catplot(df, kind<span class="op">=</span><span class="st">"strip"</span>, height<span class="op">=</span><span class="dv">3</span>, aspect<span class="op">=</span><span class="fl">1.6</span>)</span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Add a boxplot on top of the strip plot to show distribution statistics</span></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a><span class="co"># - width=0.25: makes the boxes narrower than default</span></span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a><span class="co"># - fill=False: creates transparent boxes so we can still see the strip plot points</span></span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a><span class="co"># - color="#555555": sets the box and whisker color to a dark gray</span></span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>sns.boxplot(df, width<span class="op">=</span><span class="fl">0.25</span>, fill<span class="op">=</span><span class="va">False</span>, color<span class="op">=</span><span class="st">"#555555"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="08_stats_models_files/figure-html/cell-9-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Looks like a bit of difference there, but not too strong. Time for a statistical test!</p>
<div id="f2ddbfe3" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run the Wilcoxon signed-rank test</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> stats.wilcoxon(df[<span class="st">"Pre-Treatment"</span>], df[<span class="st">"Post-Treatment"</span>]).pvalue</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co"># This is to make a nice table with the medians.</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>medians <span class="op">=</span> pd.DataFrame(df.median(), columns<span class="op">=</span>[<span class="st">"Median"</span>])</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>display(medians)</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"Mann-Whitney test p-value: </span><span class="sc">{</span>result<span class="sc">:.2f}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Median</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">Pre-Treatment</td>
<td>7.048829</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">Post-Treatment</td>
<td>8.950781</td>
</tr>
</tbody>
</table>

</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Mann-Whitney test p-value: 0.16</code></pre>
</div>
</div>
<p>In this case we fail to reject the null hypothesis that the paired samples com from the same distribution (i.e., that the difference between them is symmetric about zero).</p>
<p>Similar to the Mann-Whitney test, though the Wilcoxon signed-rank test doesn’t assume your data comes from a normal population it does s other assumptions to be aware of. Always consult the relevant statistical literature prior to using these tests in your research.</p>
</section>
</section>
</section>
<section id="comparing-multiple-groups" class="level3">
<h3 class="anchored" data-anchor-id="comparing-multiple-groups">Comparing Multiple Groups</h3>
<p>In the previous sections, we have been focused on comparing two groups. You will often run into cases in which you have more than two groups, or other experimental designs which preclude you from using more straightforward techniques. Let’s see how we can deal with that now.</p>
<p>To compare multiple groups, we most commonly use an Analysis of Variance (ANOVA) test. The <a href="https://www.jmp.com/en/statistics-knowledge-portal/one-way-anova">one-way ANOVA</a> tests the null hypothesis that two or more groups have the same population mean. It’s nonparametric partner is the <a href="https://library.virginia.edu/data/articles/getting-started-with-the-kruskal-wallis-test">Kruskal-Wallis test</a>, which uses ranks instead of raw values, and tests the null hypothesis that the population median of all of the groups are equal.</p>
<p>Let’s revisit the cancer example from last chapter <a href="07_exploratory_data_analysis.html" class="quarto-xref"><span>Chapter 7</span></a>. For the purposes here, we will generate some fake numbers for deaths per 100k people for each of the <a href="https://www.bls.gov/regions/home.htm">geographic regions</a> specified by the U.S. Bureau of Labor Statistics.</p>
<p>Imagine we have data points for three locations per state within each region (just assume these are some relevant locations for public health: counties, metro areas, whatever). In this example, we are counting the District of Columbia as a state.</p>
<p>Let’s write a function to generate the data:</p>
<div id="e0f4a88a" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Dictionary mapping each U.S. region to the number of states it contains</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>REGIONS <span class="op">=</span> {</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Northeast"</span>: <span class="dv">7</span>,</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Mid-Atlantic"</span>: <span class="dv">7</span>,</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Southeast"</span>: <span class="dv">8</span>,</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Midwest"</span>: <span class="dv">10</span>,</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Mountain-Plains"</span>: <span class="dv">6</span>,</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Southwest"</span>: <span class="dv">5</span>,</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"West"</span>: <span class="dv">8</span>,</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of sampling locations we'll collect data from in each state</span></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>SAMPLING_LOCATIONS_PER_STATE <span class="op">=</span> <span class="dv">3</span></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> generate_cancer_data(region, mean, sd):</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a><span class="co">    Generate simulated cancer death data for a specific region.</span></span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a><span class="co">    Parameters:</span></span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a><span class="co">        region (str): The name of the region from the REGIONS dictionary</span></span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a><span class="co">        mean (float): The mean cancer death count for this region</span></span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a><span class="co">        sd (float): The standard deviation of cancer deaths for this region</span></span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a><span class="co">        list: A list of dictionaries containing region and death count data</span></span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Calculate how many states are in this region</span></span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true" tabindex="-1"></a>    state_count <span class="op">=</span> REGIONS[region]</span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Calculate total number of observations needed</span></span>
<span id="cb14-32"><a href="#cb14-32" aria-hidden="true" tabindex="-1"></a>    <span class="co"># (states in region * sampling locations per state)</span></span>
<span id="cb14-33"><a href="#cb14-33" aria-hidden="true" tabindex="-1"></a>    observation_count <span class="op">=</span> state_count <span class="op">*</span> SAMPLING_LOCATIONS_PER_STATE</span>
<span id="cb14-34"><a href="#cb14-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-35"><a href="#cb14-35" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Generate random cancer death counts from a normal distribution</span></span>
<span id="cb14-36"><a href="#cb14-36" aria-hidden="true" tabindex="-1"></a>    <span class="co"># with the specified mean and standard deviation</span></span>
<span id="cb14-37"><a href="#cb14-37" aria-hidden="true" tabindex="-1"></a>    observations <span class="op">=</span> stats.norm(loc<span class="op">=</span>mean, scale<span class="op">=</span>sd).rvs(observation_count)</span>
<span id="cb14-38"><a href="#cb14-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-39"><a href="#cb14-39" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Convert the raw numbers into a list of dictionaries with region labels</span></span>
<span id="cb14-40"><a href="#cb14-40" aria-hidden="true" tabindex="-1"></a>    result <span class="op">=</span> [{<span class="st">"Region"</span>: region, <span class="st">"Deaths"</span>: observation} <span class="cf">for</span> observation <span class="kw">in</span> observations]</span>
<span id="cb14-41"><a href="#cb14-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-42"><a href="#cb14-42" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> result</span>
<span id="cb14-43"><a href="#cb14-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-44"><a href="#cb14-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-45"><a href="#cb14-45" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a pandas DataFrame by combining cancer data from all regions</span></span>
<span id="cb14-46"><a href="#cb14-46" aria-hidden="true" tabindex="-1"></a><span class="co"># Each region is assigned different mean death rates to simulate regional variations</span></span>
<span id="cb14-47"><a href="#cb14-47" aria-hidden="true" tabindex="-1"></a><span class="co"># The standard deviation is kept constant at 20 across all regions</span></span>
<span id="cb14-48"><a href="#cb14-48" aria-hidden="true" tabindex="-1"></a>cancer <span class="op">=</span> pd.DataFrame(</span>
<span id="cb14-49"><a href="#cb14-49" aria-hidden="true" tabindex="-1"></a>    generate_cancer_data(<span class="st">"Northeast"</span>, mean<span class="op">=</span><span class="dv">200</span>, sd<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb14-50"><a href="#cb14-50" aria-hidden="true" tabindex="-1"></a>    <span class="op">+</span> generate_cancer_data(<span class="st">"Mid-Atlantic"</span>, mean<span class="op">=</span><span class="dv">200</span>, sd<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb14-51"><a href="#cb14-51" aria-hidden="true" tabindex="-1"></a>    <span class="op">+</span> generate_cancer_data(<span class="st">"Southeast"</span>, mean<span class="op">=</span><span class="dv">225</span>, sd<span class="op">=</span><span class="dv">20</span>)  <span class="co"># Highest mean death rate</span></span>
<span id="cb14-52"><a href="#cb14-52" aria-hidden="true" tabindex="-1"></a>    <span class="op">+</span> generate_cancer_data(<span class="st">"Midwest"</span>, mean<span class="op">=</span><span class="dv">195</span>, sd<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb14-53"><a href="#cb14-53" aria-hidden="true" tabindex="-1"></a>    <span class="op">+</span> generate_cancer_data(<span class="st">"Mountain-Plains"</span>, mean<span class="op">=</span><span class="dv">175</span>, sd<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb14-54"><a href="#cb14-54" aria-hidden="true" tabindex="-1"></a>    <span class="op">+</span> generate_cancer_data(<span class="st">"Southwest"</span>, mean<span class="op">=</span><span class="dv">175</span>, sd<span class="op">=</span><span class="dv">20</span>)</span>
<span id="cb14-55"><a href="#cb14-55" aria-hidden="true" tabindex="-1"></a>    <span class="op">+</span> generate_cancer_data(<span class="st">"West"</span>, mean<span class="op">=</span><span class="dv">150</span>, sd<span class="op">=</span><span class="dv">20</span>)  <span class="co"># Lowest mean death rate</span></span>
<span id="cb14-56"><a href="#cb14-56" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-57"><a href="#cb14-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-58"><a href="#cb14-58" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the resulting DataFrame</span></span>
<span id="cb14-59"><a href="#cb14-59" aria-hidden="true" tabindex="-1"></a>cancer</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="10">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Region</th>
<th data-quarto-table-cell-role="th">Deaths</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>Northeast</td>
<td>232.569070</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>Northeast</td>
<td>171.141188</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>Northeast</td>
<td>196.646856</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>Northeast</td>
<td>185.546208</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>Northeast</td>
<td>215.395855</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">148</td>
<td>West</td>
<td>187.082183</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">149</td>
<td>West</td>
<td>164.184395</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">150</td>
<td>West</td>
<td>123.225502</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">151</td>
<td>West</td>
<td>199.988325</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">152</td>
<td>West</td>
<td>159.141206</td>
</tr>
</tbody>
</table>

<p>153 rows × 2 columns</p>
</div>
</div>
</div>
<p>Check out the resulting data table. We generate the data to look like this because it is a pretty good representation of how real data might be structured in a CSV or other document. Now let’s plot it:</p>
<div id="1378d3a4" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>sns.catplot(cancer, x<span class="op">=</span><span class="st">"Deaths"</span>, y<span class="op">=</span><span class="st">"Region"</span>, hue<span class="op">=</span><span class="st">"Region"</span>, kind<span class="op">=</span><span class="st">"box"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="08_stats_models_files/figure-html/cell-12-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Looks like there may be some significant differences there. Let’s run some statistical tests to see if we can assume that some of the means are different from the others.</p>
<section id="one-way-anova" class="level5">
<h5 class="anchored" data-anchor-id="one-way-anova">One-Way ANOVA</h5>
<p>We can use SciPy’s <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.f_oneway.html#scipy.stats.f_oneway">f_oneway()</a> function to run the one-way ANOVA. Here is the blurb from the docs about <code>f_oneway()</code>:</p>
<blockquote class="blockquote">
<p>The one-way ANOVA tests the null hypothesis that two or more groups have the same population mean. The test is applied to samples from two or more groups, possibly with differing sizes.</p>
</blockquote>
<p>The <code>f_oneway()</code> function doesn’t really want to work with data in the format we have it, so we have to pull out each group and pass it in separately:</p>
<div id="ca566432" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a list of Series objects, where each Series contains the 'Deaths' values</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="co"># for a specific region from the cancer dataset</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>deaths <span class="op">=</span> [cancer.query(<span class="ss">f"Region == '</span><span class="sc">{</span>region<span class="sc">}</span><span class="ss">'"</span>)[<span class="st">"Deaths"</span>] <span class="cf">for</span> region <span class="kw">in</span> REGIONS.keys()]</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform a one-way ANOVA test using the deaths data from all regions</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="co"># - The * operator unpacks the deaths list so each Series becomes a separate argument</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="co"># - stats.f_oneway compares means of two or more independent samples to determine if they're significantly different</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> stats.f_oneway(<span class="op">*</span>deaths)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the results of the ANOVA test</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a><span class="co"># - The :.1f formats the F statistic to 1 decimal place</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a><span class="co"># - The :.1e formats the p-value in scientific notation with 1 decimal place</span></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"F stat: </span><span class="sc">{</span>result<span class="sc">.</span>statistic<span class="sc">:.1f}</span><span class="ss">, p-value: </span><span class="sc">{</span>result<span class="sc">.</span>pvalue<span class="sc">:.1e}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>F stat: 21.4, p-value: 6.2e-18</code></pre>
</div>
</div>
<p>That’s a significant value, so we reject the null hypothesis that the groups have the same population mean. (Like all the other tests mentioned, it’s a good idea to make sure you are familiar with any assumptions of the test, to ensure you are using and interpreting the results correctly.)</p>
<p>Just for fun, let’s write some “fancy” code:</p>
<div id="20bb7de5" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Reshape the cancer data frame from long to wide format, with regions as columns</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>cancer_wide <span class="op">=</span> cancer.pivot(columns<span class="op">=</span><span class="st">"Region"</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a list of Series, each containing the cancer deaths for one region</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a><span class="co"># The dropna() removes any missing values from each region's data</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>deaths_by_region <span class="op">=</span> [cancer_wide[region].dropna() <span class="cf">for</span> region <span class="kw">in</span> cancer_wide.columns]</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform a one-way ANOVA test to compare means across all regions</span></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a><span class="co"># - The * operator unpacks the list so each region's data is passed as a separate argument</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> stats.f_oneway(<span class="op">*</span>deaths_by_region)</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the results, formatting F-statistic to 1 decimal place and p-value in scientific notation</span></span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"F stat: </span><span class="sc">{</span>result<span class="sc">.</span>statistic<span class="sc">:.1f}</span><span class="ss">, p-value: </span><span class="sc">{</span>result<span class="sc">.</span>pvalue<span class="sc">:.1e}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>F stat: 21.4, p-value: 6.2e-18</code></pre>
</div>
</div>
<p>Pretty neat!</p>
<div id="tip-08-which-anova" class="callout callout-style-default callout-tip callout-titled" title="Stop &amp; Think">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-5-contents" aria-controls="callout-5" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip&nbsp;8.5: Stop &amp; Think
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-5" class="callout-5-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>Which version of the code do you like better? Which do you think is more readable?</p>
</div>
</div>
</div>
</section>
<section id="kruskal-wallis" class="level5">
<h5 class="anchored" data-anchor-id="kruskal-wallis">Kruskal-Wallis</h5>
<p>The ANOVA’s non-parametric partner is the <a href="https://library.virginia.edu/data/articles/getting-started-with-the-kruskal-wallis-test">Kruskal-Wallis test</a>. It uses ranks instead of raw values, and tests the null hypothesis that the population median of all of the groups are equal.</p>
<p>The SciPy function for running the Kruskal-Wallis test is called <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.kruskal.html#scipy.stats.kruskal">kruskal()</a>. The blurb from the SciPy docs about the test is pretty informative:</p>
<blockquote class="blockquote">
<p>The Kruskal-Wallis H-test tests the null hypothesis that the population median of all of the groups are equal. It is a non-parametric version of ANOVA. The test works on 2 or more independent samples, which may have different sizes. Note that rejecting the null hypothesis does not indicate which of the groups differs. Post hoc comparisons between groups are required to determine which groups are different.</p>
</blockquote>
<p>The format is pretty much the same as for the ANOVA.</p>
<div id="0251c000" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> stats.kruskal(<span class="op">*</span>deaths)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f"H stat: </span><span class="sc">{</span>result<span class="sc">.</span>statistic<span class="sc">:.1f}</span><span class="ss">, p-value: </span><span class="sc">{</span>result<span class="sc">.</span>pvalue<span class="sc">:.1e}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>H stat: 72.5, p-value: 1.3e-13</code></pre>
</div>
</div>
<p>The p-value is less than our critical threshold of 0.05, so we reject the null hypothesis that median of all of the groups are equal. Just like with the ANOVA, it doesn’t tell us which of the groups is different. We will tackle that question in the next section.</p>
<p>I’m going to start to sound like a broken record here, but, like every other test, make sure you know the assumptions and are using the test correctly if you are using it in your own research!</p>
</section>
<section id="post-hoc-testing" class="level4">
<h4 class="anchored" data-anchor-id="post-hoc-testing">Post-Hoc Testing</h4>
<p>An ANOVA or Kruskal-Wallis test will tell you if there is a significant difference somewhere among multiple groups, but they don’t say which specific groups differ from each other. For this, we need to perform a follow up or <a href="https://statisticsbyjim.com/anova/post-hoc-tests-anova/"><em>post-hoc test</em></a>. <a href="https://www.statisticshowto.com/probability-and-statistics/statistics-definitions/post-hoc/tukey-test-honest-significant-difference/">Tukey’s Honestly Significant Difference (Tukey’s HSD)</a> is one of the most popular post-hoc methods for ANOVA (<a href="https://journals.sagepub.com/doi/pdf/10.1177/1536867X1501500117">Dunn’s test</a>, which we won’t cover in this chapter, is the most popular for Kruskal-Wallis). It systematically compares all pairs of groups and adjusts the resulting p-values (more on this in the next section). By using a post-hoc test, we can identify which groups differ significantly from each other, transforming a general “something’s different” conclusion into specific insights about where those differences lie.</p>
<p>We can run Tukey’s HSD using the <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.tukey_hsd.html#scipy.stats.tukey_hsd">tukey_hsd()</a> from the SciPy package. Here is SciPy’s docs have to say about the function:</p>
<blockquote class="blockquote">
<p>The null hypothesis is that the distributions underlying the samples all have the same mean. The test statistic, which is computed for every possible pairing of samples, is simply the difference between the sample means. For each pair, the p-value is the probability under the null hypothesis (and other assumptions; see notes) of observing such an extreme value of the statistic, considering that many pairwise comparisons are being performed. Confidence intervals for the difference between each pair of means are also available.</p>
</blockquote>
<p><em>Note: See how they mention being aware of the assumptions of the test? Make sure to check those out before using Tukey’s HSD in your own research!</em></p>
<section id="tukeys-hsd-with-scipy" class="level5">
<h5 class="anchored" data-anchor-id="tukeys-hsd-with-scipy">Tukey’s HSD with SciPy</h5>
<p>You call the <code>tukey_hsd()</code> function just like the <code>f_oneway()</code> and <code>kruskal()</code> functions. We will use the same data again:</p>
<div id="5f7c92b8" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> stats.tukey_hsd(<span class="op">*</span>deaths)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(result)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Tukey's HSD Pairwise Group Comparisons (95.0% Confidence Interval)
Comparison  Statistic  p-value  Lower CI  Upper CI
 (0 - 1)      2.357     1.000   -16.875    21.588
 (0 - 2)    -16.746     0.108   -35.367     1.875
 (0 - 3)      1.341     1.000   -16.390    19.071
 (0 - 4)     28.182     0.001     8.165    48.199
 (0 - 5)     26.825     0.004     5.757    47.892
 (0 - 6)     40.853     0.000    22.232    59.474
 (1 - 0)     -2.357     1.000   -21.588    16.875
 (1 - 2)    -19.103     0.040   -37.724    -0.482
 (1 - 3)     -1.016     1.000   -18.747    16.715
 (1 - 4)     25.825     0.003     5.808    45.842
 (1 - 5)     24.468     0.012     3.401    45.535
 (1 - 6)     38.496     0.000    19.875    57.117
 (2 - 0)     16.746     0.108    -1.875    35.367
 (2 - 1)     19.103     0.040     0.482    37.724
 (2 - 3)     18.087     0.030     1.021    35.153
 (2 - 4)     44.928     0.000    25.497    64.359
 (2 - 5)     43.571     0.000    23.060    64.082
 (2 - 6)     57.599     0.000    39.609    75.588
 (3 - 0)     -1.341     1.000   -19.071    16.390
 (3 - 1)      1.016     1.000   -16.715    18.747
 (3 - 2)    -18.087     0.030   -35.153    -1.021
 (3 - 4)     26.841     0.001     8.261    45.420
 (3 - 5)     25.484     0.003     5.777    45.190
 (3 - 6)     39.512     0.000    22.446    56.578
 (4 - 0)    -28.182     0.001   -48.199    -8.165
 (4 - 1)    -25.825     0.003   -45.842    -5.808
 (4 - 2)    -44.928     0.000   -64.359   -25.497
 (4 - 3)    -26.841     0.001   -45.420    -8.261
 (4 - 5)     -1.357     1.000   -23.143    20.429
 (4 - 6)     12.671     0.451    -6.760    32.102
 (5 - 0)    -26.825     0.004   -47.892    -5.757
 (5 - 1)    -24.468     0.012   -45.535    -3.401
 (5 - 2)    -43.571     0.000   -64.082   -23.060
 (5 - 3)    -25.484     0.003   -45.190    -5.777
 (5 - 4)      1.357     1.000   -20.429    23.143
 (5 - 6)     14.028     0.391    -6.483    34.539
 (6 - 0)    -40.853     0.000   -59.474   -22.232
 (6 - 1)    -38.496     0.000   -57.117   -19.875
 (6 - 2)    -57.599     0.000   -75.588   -39.609
 (6 - 3)    -39.512     0.000   -56.578   -22.446
 (6 - 4)    -12.671     0.451   -32.102     6.760
 (6 - 5)    -14.028     0.391   -34.539     6.483
</code></pre>
</div>
</div>
<p>The <code>tukey_hsd()</code> function returns an instance of the <code>TukeyHSDResult</code> class. You may find that the <code>TukeyHSDResult</code> class is not exactly the most user-friendly class. To get the data to put on a chart, you have to do a little bit of work:</p>
<div id="54241d60" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform Tukey's HSD (Honest Significant Difference) test on the death data</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="co"># This test compares all possible pairs of regions to find significant differences</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> stats.tukey_hsd(<span class="op">*</span>deaths)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate 95% confidence intervals for the differences between means</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>confidence_interval <span class="op">=</span> result.confidence_interval(<span class="fl">0.95</span>)</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the list of region names from the REGIONS dictionary</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>regions <span class="op">=</span> <span class="bu">list</span>(REGIONS.keys())</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Create an empty list to store the results for each pair of regions</span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>rows <span class="op">=</span> []</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through all pairs of regions using numpy's ndenumerate to get indices and p-values</span></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i, j), p_value <span class="kw">in</span> np.ndenumerate(result.pvalue):</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Skip comparing a region to itself (i.e., the diagonal elements)</span></span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">!=</span> j:</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Create a dictionary with all the information for this pair of regions</span></span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>        row <span class="op">=</span> {</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Region 1"</span>: regions[i],</span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Region 2"</span>: regions[j],</span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Create a label for the pair</span></span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Pair"</span>: regions[i] <span class="op">+</span> <span class="st">"-"</span> <span class="op">+</span> regions[j],</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>            <span class="co"># The mean difference between regions</span></span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Difference in Means"</span>: result.statistic[i, j],</span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a>            <span class="st">"p-value"</span>: p_value,</span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a>            <span class="st">"Significant?"</span>: p_value <span class="op">&lt;</span> <span class="fl">0.05</span>,</span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Lower bound of confidence interval</span></span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a>            <span class="st">"CI Low"</span>: confidence_interval.low[i, j],</span>
<span id="cb24-30"><a href="#cb24-30" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Upper bound of confidence interval</span></span>
<span id="cb24-31"><a href="#cb24-31" aria-hidden="true" tabindex="-1"></a>            <span class="st">"CI High"</span>: confidence_interval.high[i, j],</span>
<span id="cb24-32"><a href="#cb24-32" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb24-33"><a href="#cb24-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-34"><a href="#cb24-34" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Add this pair's results to our list</span></span>
<span id="cb24-35"><a href="#cb24-35" aria-hidden="true" tabindex="-1"></a>        rows.append(row)</span>
<span id="cb24-36"><a href="#cb24-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-37"><a href="#cb24-37" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert the list of dictionaries to a pandas DataFrame</span></span>
<span id="cb24-38"><a href="#cb24-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Sort by the difference in means and reset the index</span></span>
<span id="cb24-39"><a href="#cb24-39" aria-hidden="true" tabindex="-1"></a>tukey_df <span class="op">=</span> pd.DataFrame(rows).sort_values(<span class="st">"Difference in Means"</span>).reset_index()</span>
<span id="cb24-40"><a href="#cb24-40" aria-hidden="true" tabindex="-1"></a>tukey_df  <span class="co"># Display the DataFrame</span></span>
<span id="cb24-41"><a href="#cb24-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-42"><a href="#cb24-42" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a categorical plot using seaborn</span></span>
<span id="cb24-43"><a href="#cb24-43" aria-hidden="true" tabindex="-1"></a><span class="co"># This will show the difference in means for each pair of regions</span></span>
<span id="cb24-44"><a href="#cb24-44" aria-hidden="true" tabindex="-1"></a>facet_grid <span class="op">=</span> sns.catplot(</span>
<span id="cb24-45"><a href="#cb24-45" aria-hidden="true" tabindex="-1"></a>    tukey_df,</span>
<span id="cb24-46"><a href="#cb24-46" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span><span class="st">"Pair"</span>,</span>
<span id="cb24-47"><a href="#cb24-47" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span><span class="st">"Difference in Means"</span>,</span>
<span id="cb24-48"><a href="#cb24-48" aria-hidden="true" tabindex="-1"></a>    hue<span class="op">=</span><span class="st">"Significant?"</span>,</span>
<span id="cb24-49"><a href="#cb24-49" aria-hidden="true" tabindex="-1"></a>    height<span class="op">=</span><span class="dv">7</span>,</span>
<span id="cb24-50"><a href="#cb24-50" aria-hidden="true" tabindex="-1"></a>    aspect<span class="op">=</span><span class="fl">0.8</span>,</span>
<span id="cb24-51"><a href="#cb24-51" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb24-52"><a href="#cb24-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-53"><a href="#cb24-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the current axis object to add more elements to the plot</span></span>
<span id="cb24-54"><a href="#cb24-54" aria-hidden="true" tabindex="-1"></a><span class="co"># - 'ax' is a common abbreviation you will see when using matplotlib</span></span>
<span id="cb24-55"><a href="#cb24-55" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> facet_grid.facet_axis(<span class="dv">0</span>, <span class="dv">0</span>)</span>
<span id="cb24-56"><a href="#cb24-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-57"><a href="#cb24-57" aria-hidden="true" tabindex="-1"></a><span class="co"># Add confidence interval lines for each pair</span></span>
<span id="cb24-58"><a href="#cb24-58" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> idx, row <span class="kw">in</span> tukey_df.iterrows():</span>
<span id="cb24-59"><a href="#cb24-59" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Plot a horizontal line for each confidence interval</span></span>
<span id="cb24-60"><a href="#cb24-60" aria-hidden="true" tabindex="-1"></a>    ax.plot([row[<span class="st">"CI Low"</span>], row[<span class="st">"CI High"</span>]], [idx, idx], color<span class="op">=</span><span class="st">"#333333"</span>, linewidth<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb24-61"><a href="#cb24-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-62"><a href="#cb24-62" aria-hidden="true" tabindex="-1"></a><span class="co"># Add a vertical line at x=0</span></span>
<span id="cb24-63"><a href="#cb24-63" aria-hidden="true" tabindex="-1"></a><span class="co"># This helps visualize which differences are positive vs. negative</span></span>
<span id="cb24-64"><a href="#cb24-64" aria-hidden="true" tabindex="-1"></a><span class="co"># If a confidence interval crosses this line, the difference is not significant</span></span>
<span id="cb24-65"><a href="#cb24-65" aria-hidden="true" tabindex="-1"></a>ax.axvline(<span class="dv">0</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, color<span class="op">=</span><span class="st">"#BBBBBB"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="08_stats_models_files/figure-html/cell-17-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p><em>Note: You may want to check out the docs for the <a href="https://numpy.org/doc/2.2/reference/generated/numpy.ndenumerate.html">numpy.ndenumerate()</a> function as we haven’t covered it before now.</em></p>
<p>We added in the confidence intervals to the plot. As you can see, the pairs whose confidence interval overlaps with zero are not-significantly different by the test, while those whose confidence interval does <em>not</em> overlap with zero are significantly different by the test.</p>
<p>This code is a bit finicky, and we would like to avoid that sort of code whenever possible. Additionally, the plot itself is fairly tricky to read since there are so many pairs. To help with both these issues, let’s pull in the <a href="https://www.statsmodels.org/stable/index.html">statsmodels</a> package as it has a nice, built-in way to visualize Tukey’s HSD results.</p>
</section>
<section id="tukeys-hsd-with-statsmodels" class="level5">
<h5 class="anchored" data-anchor-id="tukeys-hsd-with-statsmodels">Tukey’s HSD with statsmodels</h5>
<p>To run the test using the statsmodels package, we use <a href="https://www.statsmodels.org/dev/generated/statsmodels.stats.multicomp.pairwise_tukeyhsd.html">pairwise_tukeyhsd()</a>, and to generate the plot we use the <a href="https://www.statsmodels.org/dev/generated/statsmodels.sandbox.stats.multicomp.TukeyHSDResults.plot_simultaneous.html#statsmodels.sandbox.stats.multicomp.TukeyHSDResults.plot_simultaneous">plot_simultaneous()</a> function. Let’s see how easy it is to run:</p>
<div id="a5225fa7" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> statsmodels.stats.multicomp.pairwise_tukeyhsd(cancer[<span class="st">"Deaths"</span>], groups<span class="op">=</span>cancer[<span class="st">"Region"</span>], alpha<span class="op">=</span><span class="fl">0.05</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="co"># There is a small issue where if you don't set the result of this to a variable,</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co"># you get two charts instead of one.</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>figure <span class="op">=</span> result.plot_simultaneous()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="08_stats_models_files/figure-html/cell-18-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>That’s pretty cool! Now, instead of looking at all those pairs, we can look at confidence intervals for each region. If they do not overlap with another region, then those two regions have significantly different deaths per 100k people. Not only did we get a plot that more effectively shows the results, we also got to chop out a ton of code!</p>
<div id="tip-08-post-hoc-interpretation" class="callout callout-style-default callout-tip callout-titled" title="Stop &amp; Think">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip&nbsp;8.6: Stop &amp; Think
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>Looking at the Tukey’s HSD plot, how would you explain which regions have significantly different cancer death rates?</p>
</div>
</div>
</div>
</section>
</section>
</section>
</section>
<section id="linear-regression" class="level2">
<h2 class="anchored" data-anchor-id="linear-regression">Linear Regression</h2>
<p>Linear regression is incredibly common, and I’m absolutely positive that you all have already used it in your research. Let’s see how it goes in Python!</p>
<p>Linear regression uses one or more explanatory (independent) variables to predict a response (dependent) variable. Simple linear regression refers to a model with only one explanatory variable, while models with two or more explanatory variables are multiple linear regression models. In both cases, we fit a linear model to the data, meaning that we are expecting a straight, non-vertical line. (There’s a really snazzy visual guide to linear regression <a href="https://mlu-explain.github.io/linear-regression/">here</a>.)</p>
<p>Let’s generate and take a look at some data.</p>
<div id="9d2487a4" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This ensures that random numbers generated will be the same every time the</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># code runs</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>np.random.seed(<span class="dv">254370</span>)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Define how many data points we want to generate</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>observation_count <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate X values from a uniform distribution</span></span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a><span class="co"># - loc=-50: The lower bound of the distribution</span></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a><span class="co"># - scale=100: The range of the distribution (from -50 to -50+100=50)</span></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a><span class="co"># - size=observation_count: Generate 100 random values</span></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> stats.uniform.rvs(loc<span class="op">=-</span><span class="dv">50</span>, scale<span class="op">=</span><span class="dv">100</span>, size<span class="op">=</span>observation_count)</span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate Y values based on X following a linear relationship: Y = 2X + noise</span></span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a><span class="co"># 2 * x: This creates a linear relationship where Y increases by 2 for every</span></span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a><span class="co">#   unit increase in X</span></span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a><span class="co"># stats.norm.rvs(...): This adds random noise from a normal distribution</span></span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a><span class="co"># - loc=0: The noise is centered around 0 (no bias)</span></span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a><span class="co"># - scale=20: The standard deviation of the noise is 20</span></span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a><span class="co">#   (determines how scattered the points are)</span></span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a><span class="co"># - size=observation_count: Generate 100 random noise values</span></span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a>y <span class="op">=</span> <span class="dv">2</span> <span class="op">*</span> x <span class="op">+</span> stats.norm.rvs(loc<span class="op">=</span><span class="dv">0</span>, scale<span class="op">=</span><span class="dv">20</span>, size<span class="op">=</span>observation_count)</span>
<span id="cb26-23"><a href="#cb26-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-24"><a href="#cb26-24" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame({<span class="st">"X"</span>: x, <span class="st">"Y"</span>: y})</span>
<span id="cb26-25"><a href="#cb26-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-26"><a href="#cb26-26" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a scatter plot to visualize the relationship between X and Y</span></span>
<span id="cb26-27"><a href="#cb26-27" aria-hidden="true" tabindex="-1"></a>df.plot(kind<span class="op">=</span><span class="st">"scatter"</span>, x<span class="op">=</span><span class="st">"X"</span>, y<span class="op">=</span><span class="st">"Y"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="08_stats_models_files/figure-html/cell-19-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Looks like a pretty strong trend there. Let’s build a linear model using ordinary least squares fitting. We can use the <a href="https://www.statsmodels.org/dev/generated/statsmodels.formula.api.ols.html#statsmodels.formula.api.ols">ols()</a> function from the statsmodels package for this. This package has some neat features like the ability to specify models in a way that looks a lot like how we might do it in R.</p>
<div id="38f76127" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import necessary libraries (assumed to be done before this code)</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="co"># statsmodels is used for statistical modeling</span></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="co"># sns (seaborn) is used for data visualization</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Describe the model using the formula interface</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a><span class="co"># "Y ~ X" means we want to predict Y based on X (simple linear regression)</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="co"># This creates a model object but doesn't fit it to the data yet</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> statsmodels.formula.api.ols(<span class="st">"Y ~ X"</span>, data<span class="op">=</span>df)</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit the model to the data</span></span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a><span class="co"># This performs the actual regression calculations and returns a results object</span></span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a><span class="co"># containing coefficients, p-values, R-squared, etc.</span></span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> model.fit()</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Display a summary of the regression results</span></span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a><span class="co"># The summary() method provides a comprehensive report of the model performance</span></span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a><span class="co"># 'slim=True' parameter gives a condensed version of the output with just essential</span></span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a><span class="co">#   statistics</span></span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a><span class="co"># This will show coefficients, standard errors, t-values, p-values, R-squared, etc.</span></span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(result.summary(slim<span class="op">=</span><span class="va">True</span>))</span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a scatter plot with regression line using seaborn</span></span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a><span class="co"># lmplot automatically:</span></span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a><span class="co"># - Creates a scatter plot of the raw data (X vs Y)</span></span>
<span id="cb27-25"><a href="#cb27-25" aria-hidden="true" tabindex="-1"></a><span class="co"># - Fits a regression line</span></span>
<span id="cb27-26"><a href="#cb27-26" aria-hidden="true" tabindex="-1"></a><span class="co"># - Adds a shaded confidence interval around the line</span></span>
<span id="cb27-27"><a href="#cb27-27" aria-hidden="true" tabindex="-1"></a><span class="co"># This provides a visual representation of the relationship we just modeled</span></span>
<span id="cb27-28"><a href="#cb27-28" aria-hidden="true" tabindex="-1"></a><span class="co"># statistically</span></span>
<span id="cb27-29"><a href="#cb27-29" aria-hidden="true" tabindex="-1"></a>sns.lmplot(df, x<span class="op">=</span><span class="st">"X"</span>, y<span class="op">=</span><span class="st">"Y"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      Y   R-squared:                       0.877
Model:                            OLS   Adj. R-squared:                  0.875
No. Observations:                 100   F-statistic:                     696.5
Covariance Type:            nonrobust   Prob (F-statistic):           2.51e-46
==============================================================================
                 coef    std err          t      P&gt;|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.5727      2.091      1.230      0.221      -1.576       6.722
X              1.9263      0.073     26.391      0.000       1.781       2.071
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="08_stats_models_files/figure-html/cell-20-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The summary table gives you the main regression result info, including fitted coefficients, standard errors, p-values, R-squared values, etc. We won’t get into the details of how regression is calculated or how exactly to interpret it in this chapter, so just focus on the coefficients and the p-values for now.</p>
<p>In this case, we have a coefficient for <code>X</code> of 1.9, which is pretty close to the true model, and it is significant! Additionally, the R-squared value is 0.877, which means this model captures a lot of the variability in our data. In other words, it’s a very good model!</p>
<section id="with-no-relationship-in-the-data" class="level3">
<h3 class="anchored" data-anchor-id="with-no-relationship-in-the-data">With No Relationship in the Data</h3>
<p>Let’s try another dataset, but this time the response is not related to the predictors.</p>
<div id="64c39146" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the seed</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>np.random.seed(<span class="dv">962378</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate the data</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>observation_count <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>x <span class="op">=</span> stats.uniform(loc<span class="op">=-</span><span class="dv">50</span>, scale<span class="op">=</span><span class="dv">100</span>).rvs(size<span class="op">=</span>observation_count)</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a><span class="co"># The response values are basically random noise</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>y <span class="op">=</span> stats.norm(loc<span class="op">=</span><span class="dv">0</span>, scale<span class="op">=</span><span class="dv">10</span>).rvs(size<span class="op">=</span>observation_count)</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame({<span class="st">"X"</span>: x, <span class="st">"Y"</span>: y})</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the model</span></span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> statsmodels.formula.api.ols(<span class="st">"Y ~ X"</span>, data<span class="op">=</span>df)</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit the model</span></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> model.fit()</span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the summary and plot the data</span></span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(result.summary(slim<span class="op">=</span><span class="va">True</span>))</span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>sns.lmplot(df, x<span class="op">=</span><span class="st">"X"</span>, y<span class="op">=</span><span class="st">"Y"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      Y   R-squared:                       0.001
Model:                            OLS   Adj. R-squared:                 -0.009
No. Observations:                 100   F-statistic:                   0.08288
Covariance Type:            nonrobust   Prob (F-statistic):              0.774
==============================================================================
                 coef    std err          t      P&gt;|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.3901      1.057      0.369      0.713      -1.707       2.487
X             -0.0115      0.040     -0.288      0.774      -0.091       0.068
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="08_stats_models_files/figure-html/cell-21-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>This time, we have pretty much the opposite result. The fitted coefficient is around zero, the line on the plot is basically horizontal, and the R-squared value is very low. So it is very likely that there is no relationship that we can measure between these two variables. And of course, we expected this result, as that is how we set up the data!</p>
<div id="tip-08-regression-interpretation" class="callout callout-style-default callout-tip callout-titled" title="Stop &amp; Think">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-7-contents" aria-controls="callout-7" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip&nbsp;8.7: Stop &amp; Think
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-7" class="callout-7-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>What does an R-squared value close to zero tell you about a linear regression model?</p>
</div>
</div>
</div>
</section>
<section id="multiple-predictors" class="level3">
<h3 class="anchored" data-anchor-id="multiple-predictors">Multiple Predictors</h3>
<p>Multiple predictors can be given as well. Let’s make another data set with three predictors, <span class="math inline">\(x_1\)</span>, <span class="math inline">\(x_2\)</span>, and <span class="math inline">\(x_3\)</span>. The response <span class="math inline">\(y\)</span> will only depend on the first two predictors, however.</p>
<div id="5769055b" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>np.random.seed(<span class="dv">526347</span>)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>observation_count <span class="op">=</span> <span class="dv">100</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate 3 independent variables from uniform distributions</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Each variable ranges from -50 to 50 (loc=-50, scale=100)</span></span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>x1 <span class="op">=</span> stats.uniform(loc<span class="op">=-</span><span class="dv">50</span>, scale<span class="op">=</span><span class="dv">100</span>).rvs(size<span class="op">=</span>observation_count)</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>x2 <span class="op">=</span> stats.uniform(loc<span class="op">=-</span><span class="dv">50</span>, scale<span class="op">=</span><span class="dv">100</span>).rvs(size<span class="op">=</span>observation_count)</span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>x3 <span class="op">=</span> stats.uniform(loc<span class="op">=-</span><span class="dv">50</span>, scale<span class="op">=</span><span class="dv">100</span>).rvs(size<span class="op">=</span>observation_count)</span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate random noise from a normal distribution with mean 0 and standard</span></span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a><span class="co"># deviation 10</span></span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>noise <span class="op">=</span> stats.norm(loc<span class="op">=</span><span class="dv">0</span>, scale<span class="op">=</span><span class="dv">10</span>).rvs(size<span class="op">=</span>observation_count)</span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the dependent variable y as a linear combination of x1 and x2,</span></span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a><span class="co">#   plus noise</span></span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a><span class="co"># True model: y = 2*x1 - 3*x2 + noise</span></span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a><span class="co"># - Note that x3 is not used in generating y!</span></span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a>y <span class="op">=</span> <span class="dv">2</span> <span class="op">*</span> x1 <span class="op">-</span> <span class="dv">3</span> <span class="op">*</span> x2 <span class="op">+</span> noise</span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a DataFrame with all variables</span></span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Sort it by the y values and reset the index</span></span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> (</span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a>    pd.DataFrame({<span class="st">"X1"</span>: x1, <span class="st">"X2"</span>: x2, <span class="st">"X3"</span>: x3, <span class="st">"Y"</span>: y})</span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a>    .sort_values(<span class="st">"Y"</span>)</span>
<span id="cb31-26"><a href="#cb31-26" aria-hidden="true" tabindex="-1"></a>    .reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb31-27"><a href="#cb31-27" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb31-28"><a href="#cb31-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-29"><a href="#cb31-29" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit a linear regression model using all three predictors</span></span>
<span id="cb31-30"><a href="#cb31-30" aria-hidden="true" tabindex="-1"></a><span class="co"># - Formula notation: Y ~ X1 + X2 + X3</span></span>
<span id="cb31-31"><a href="#cb31-31" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> statsmodels.formula.api.ols(<span class="st">"Y ~ X1 + X2 + X3"</span>, data<span class="op">=</span>df)</span>
<span id="cb31-32"><a href="#cb31-32" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> model.fit()</span>
<span id="cb31-33"><a href="#cb31-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-34"><a href="#cb31-34" aria-hidden="true" tabindex="-1"></a><span class="co"># Display a summary of the regression results</span></span>
<span id="cb31-35"><a href="#cb31-35" aria-hidden="true" tabindex="-1"></a><span class="co"># - slim=True omits some of the more detailed statistics</span></span>
<span id="cb31-36"><a href="#cb31-36" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(result.summary(slim<span class="op">=</span><span class="va">True</span>))</span>
<span id="cb31-37"><a href="#cb31-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-38"><a href="#cb31-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a pairplot showing relationships between Y and each predictor</span></span>
<span id="cb31-39"><a href="#cb31-39" aria-hidden="true" tabindex="-1"></a><span class="co"># - kind="reg" adds regression lines to the scatter plots</span></span>
<span id="cb31-40"><a href="#cb31-40" aria-hidden="true" tabindex="-1"></a>sns.pairplot(df, kind<span class="op">=</span><span class="st">"reg"</span>, y_vars<span class="op">=</span><span class="st">"Y"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      Y   R-squared:                       0.990
Model:                            OLS   Adj. R-squared:                  0.989
No. Observations:                 100   F-statistic:                     3044.
Covariance Type:            nonrobust   Prob (F-statistic):           5.20e-95
==============================================================================
                 coef    std err          t      P&gt;|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.1886      0.986      1.205      0.231      -0.769       3.146
X1             2.0538      0.033     62.303      0.000       1.988       2.119
X2            -2.9983      0.037    -81.967      0.000      -3.071      -2.926
X3             0.0445      0.034      1.318      0.191      -0.023       0.112
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="08_stats_models_files/figure-html/cell-22-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The coefficients of the variables, p-values, and scatter plots are all about what we would expect given the way we generated the data.</p>
<p>One cool thing to call out is the <a href="https://seaborn.pydata.org/generated/seaborn.pairplot.html">pairplot()</a> function from seaborn. It’s a nice way to visualize multiple variables at the same time. Here, we can see a positive association between <code>X1</code> and <code>y</code>, a negative association between <code>X2</code> and <code>y</code>, and no association between <code>X3</code> and <code>y</code>. Cool!</p>
<p>Linear regression assumes some things about your data (are you tired of me saying this yet…). For example, the <a href="https://www.jmp.com/en/statistics-knowledge-portal/what-is-regression/simple-linear-regression-assumptions">JMP docs</a> put it this way:</p>
<blockquote class="blockquote">
<ul>
<li>The true relationship is linear</li>
<li>Errors are normally distributed</li>
<li>Homoscedasticity of errors (or, equal variance around the line).</li>
<li>Independence of the observations</li>
</ul>
</blockquote>
<p>It’s a good idea to check these assumptions before interpreting your results. There are many common <a href="https://library.virginia.edu/data/articles/diagnostic-plots">diagnostic plots</a> for doing so, like the normal Q-Q plot, and the scale-location plot, but that is a story for another time!</p>
<p><em>Note: There is a really cool site that talks about how most <a href="https://lindeloev.github.io/tests-as-linear/">common statistical models are basically linear models</a> (or at least, very close to them). It’s a pretty neat way to think of statistical modeling that I encourage you to check out!</em></p>
<div id="tip-08-feature-selection" class="callout callout-style-default callout-tip callout-titled" title="Stop &amp; Think">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip&nbsp;8.8: Stop &amp; Think
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>In our multiple regression example, how could you determine which predictors are most important for the model?</p>
</div>
</div>
</div>
</section>
</section>
<section id="dimensionality-reduction-clustering" class="level2">
<h2 class="anchored" data-anchor-id="dimensionality-reduction-clustering">Dimensionality Reduction &amp; Clustering</h2>
<p>Let’s switch gears a little bit and introduce dimensionality reduction and clustering. Dimensionality reduction techniques help transform complex high-dimensional data (like genomic or proteomic datasets with thousands of features) into simpler representations that capture the essential patterns. Clustering methods then help identify natural groupings within this data, allowing you to discover hidden structures without prior labeling. These techniques are common tools in bioinformatics that allow researchers to visualize complex biological relationships, identify subtypes of diseases, or group similar protein structures together when manual classification would be impractical.</p>
<section id="principal-components-analysis-pca" class="level3">
<h3 class="anchored" data-anchor-id="principal-components-analysis-pca">Principal Components Analysis (PCA)</h3>
<p>Principal Component Analysis (PCA) is a widely used dimensionality reduction technique that transforms your original variables into a new set of uncorrelated variables called principal components. These principal components are ordered by how much of the original data’s variance they capture, with the first component accounting for the most variance, the second component capturing the second most, and so on. PCA is more or less finding the directions (or axes) along which your data varies the most, allowing you to reduce dimensionality while preserving as much information as possible.</p>
<p><em>Note: there is another way to think of PCA, and that is as a form of <a href="https://ordination.okstate.edu/PCA.htm">indirect gradient analysis</a> or as <a href="https://doi.org/10.1111/1467-9876.00275">biplots of compositional data</a>, which are two of my favorite topics. Unfortunately, or maybe fortunately depending on your interests, we won’t be covering that here!</em></p>
<p>First, let’s read in the data. You might recognize this dataset, its Fisher’s famous iris dataset!</p>
<div id="9f874db0" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>iris <span class="op">=</span> pd.read_csv(<span class="st">"../_data/iris.csv"</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>display(iris)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>sns.pairplot(iris, hue<span class="op">=</span><span class="st">"Species"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Sepal.Length</th>
<th data-quarto-table-cell-role="th">Sepal.Width</th>
<th data-quarto-table-cell-role="th">Petal.Length</th>
<th data-quarto-table-cell-role="th">Petal.Width</th>
<th data-quarto-table-cell-role="th">Species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>5.1</td>
<td>3.5</td>
<td>1.4</td>
<td>0.2</td>
<td>setosa</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>4.9</td>
<td>3.0</td>
<td>1.4</td>
<td>0.2</td>
<td>setosa</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>4.7</td>
<td>3.2</td>
<td>1.3</td>
<td>0.2</td>
<td>setosa</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>4.6</td>
<td>3.1</td>
<td>1.5</td>
<td>0.2</td>
<td>setosa</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>5.0</td>
<td>3.6</td>
<td>1.4</td>
<td>0.2</td>
<td>setosa</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">145</td>
<td>6.7</td>
<td>3.0</td>
<td>5.2</td>
<td>2.3</td>
<td>virginica</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">146</td>
<td>6.3</td>
<td>2.5</td>
<td>5.0</td>
<td>1.9</td>
<td>virginica</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">147</td>
<td>6.5</td>
<td>3.0</td>
<td>5.2</td>
<td>2.0</td>
<td>virginica</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">148</td>
<td>6.2</td>
<td>3.4</td>
<td>5.4</td>
<td>2.3</td>
<td>virginica</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">149</td>
<td>5.9</td>
<td>3.0</td>
<td>5.1</td>
<td>1.8</td>
<td>virginica</td>
</tr>
</tbody>
</table>

<p>150 rows × 5 columns</p>
</div>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="08_stats_models_files/figure-html/cell-23-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can use the <a href="https://www.statsmodels.org/stable/generated/statsmodels.multivariate.pca.PCA.html#statsmodels.multivariate.pca.PCA">PCA</a> class from the statsmodels package to compute the PCA:</p>
<div id="ea080d76" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>result <span class="op">=</span> PCA(</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>    <span class="co"># The data to perform PCA on - our iris dataset without the 'Species' column</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>    <span class="co"># We're only using the numeric feature columns</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>    iris.drop(columns<span class="op">=</span><span class="st">"Species"</span>),</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># demean=True means that the mean will be subtracted from each feature</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># This centers the data around zero</span></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># R's `prcomp()` calls this parameter "center"</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>    demean<span class="op">=</span><span class="va">True</span>,</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># standardize=False means that we are NOT scaling each feature to have unit variance</span></span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If True, each feature would be divided by its standard deviation</span></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># R's `prcomp()` would use both center = TRUE and scale. = TRUE for standardization</span></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>    standardize<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>    <span class="co"># We won't get into the math behind this parameter.  Just know that setting it to</span></span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>    <span class="co"># False is the more typical use case for our purposes.</span></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>    normalize<span class="op">=</span><span class="va">False</span>,</span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># ncomp=2 specifies that we want to keep only the first 2 principal components</span></span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>    <span class="co"># These will be the 2 directions that capture the most variance in the data</span></span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a>    ncomp<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a>result</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="23">
<pre><code>Principal Component Analysis(nobs: 150, nvar: 4, transformation: Demean (Covariance), normalization: False, number of components: 2, SVD, id: 0x120c2b0e0)</code></pre>
</div>
</div>
<p>Once PCA is performed, we can visualize the data by plotting the projections onto the first two principal components:</p>
<div id="e369829f" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Prepare the data for plotting</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>plot_data <span class="op">=</span> (</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>    result.scores.rename(</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>        columns<span class="op">=</span>{</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Rename the first principal component for better readability</span></span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>            <span class="st">"comp_0"</span>: <span class="st">"PC 1"</span>,</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Rename the second principal component for better readability</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>            <span class="st">"comp_1"</span>: <span class="st">"PC 2"</span>,</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Add the Species column from the original dataset for coloring points</span></span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>    .assign(Species<span class="op">=</span>iris[<span class="st">"Species"</span>])</span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a scatter plot using seaborn</span></span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a><span class="co"># - relplot creates a relational plot (scatter plot in this case)</span></span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a><span class="co"># - x and y specify which columns to use for the x and y axes</span></span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a><span class="co">#   (the first two principal components)</span></span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a><span class="co"># - hue colors the points based on the Species column</span></span>
<span id="cb36-20"><a href="#cb36-20" aria-hidden="true" tabindex="-1"></a>sns.relplot(plot_data, x<span class="op">=</span><span class="st">"PC 1"</span>, y<span class="op">=</span><span class="st">"PC 2"</span>, hue<span class="op">=</span><span class="st">"Species"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="08_stats_models_files/figure-html/cell-25-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>As you can see, the three species are fairly well separated along the x-axis, which represents the first principal component.</p>
<div id="tip-08-pca-interpretation" class="callout callout-style-default callout-tip callout-titled" title="Stop &amp; Think">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-9-contents" aria-controls="callout-9" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip&nbsp;8.9: Stop &amp; Think
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-9" class="callout-9-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>What biological interpretation might you give to the first principal component in this iris dataset analysis?</p>
</div>
</div>
</div>
<p>There is a lot more to say about PCA, including differences between using the covariance vs.&nbsp;correlation matrix, biplots, interpretation, etc., but we will leave it there for now.</p>
</section>
<section id="k-means-clustering" class="level3">
<h3 class="anchored" data-anchor-id="k-means-clustering">k-means Clustering</h3>
<p>K-means clustering is a common unsupervised machine learning technique that partitions your data into a specified number, <code>k</code>, of distinct, non-overlapping groups, without requiring labeled training data. The algorithm works by iteratively assigning data points to the nearest cluster center, then recalculating those centers until convergence is reached. Some use cases for k-means clustering include grouping genes given expression data, identifying protein function similarities, or finding patient subgroups with similar clinical profiles.</p>
<p>We can use SciPy’s <a href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.vq.kmeans2.html">kmeans2()</a> function to run the k-means algorithm on our iris data:</p>
<div id="2a5a4abd" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the random seed for reproducibility</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>np.random.seed(<span class="dv">238974</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform K-means clustering on the iris dataset</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="co"># - We're using the numeric columns of the iris data frame</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a><span class="co"># - We specify k=3 to create 3 clusters (matching the number of iris species)</span></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a><span class="co"># - We ignore the _centroids</span></span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a><span class="co"># - `labels` will contain the cluster assignment (0, 1, or 2) for each iris sample</span></span>
<span id="cb37-9"><a href="#cb37-9" aria-hidden="true" tabindex="-1"></a>_centroids, labels <span class="op">=</span> cluster.vq.kmeans2(iris.drop(columns<span class="op">=</span><span class="st">"Species"</span>), k<span class="op">=</span><span class="dv">3</span>)</span>
<span id="cb37-10"><a href="#cb37-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-11"><a href="#cb37-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a scatter plot to visualize the clustering results</span></span>
<span id="cb37-12"><a href="#cb37-12" aria-hidden="true" tabindex="-1"></a><span class="co"># - We're plotting the first two principal components (PC 1 vs PC 2)</span></span>
<span id="cb37-13"><a href="#cb37-13" aria-hidden="true" tabindex="-1"></a><span class="co"># - Each point is colored by its actual species (using 'hue')</span></span>
<span id="cb37-14"><a href="#cb37-14" aria-hidden="true" tabindex="-1"></a><span class="co"># - Each point's marker style is determined by its cluster assignment (using 'style')</span></span>
<span id="cb37-15"><a href="#cb37-15" aria-hidden="true" tabindex="-1"></a><span class="co"># - This allows us to compare the actual species classifications with the clustering</span></span>
<span id="cb37-16"><a href="#cb37-16" aria-hidden="true" tabindex="-1"></a><span class="co">#   results</span></span>
<span id="cb37-17"><a href="#cb37-17" aria-hidden="true" tabindex="-1"></a><span class="co"># - 'plot_data.assign(Cluster=labels)' adds the cluster labels as a new column to the</span></span>
<span id="cb37-18"><a href="#cb37-18" aria-hidden="true" tabindex="-1"></a><span class="co">#   data frame</span></span>
<span id="cb37-19"><a href="#cb37-19" aria-hidden="true" tabindex="-1"></a>sns.relplot(</span>
<span id="cb37-20"><a href="#cb37-20" aria-hidden="true" tabindex="-1"></a>    plot_data.assign(Cluster<span class="op">=</span>labels),</span>
<span id="cb37-21"><a href="#cb37-21" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span><span class="st">"PC 1"</span>,</span>
<span id="cb37-22"><a href="#cb37-22" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span><span class="st">"PC 2"</span>,</span>
<span id="cb37-23"><a href="#cb37-23" aria-hidden="true" tabindex="-1"></a>    hue<span class="op">=</span><span class="st">"Species"</span>,</span>
<span id="cb37-24"><a href="#cb37-24" aria-hidden="true" tabindex="-1"></a>    style<span class="op">=</span><span class="st">"Cluster"</span>,</span>
<span id="cb37-25"><a href="#cb37-25" aria-hidden="true" tabindex="-1"></a>    legend<span class="op">=</span><span class="st">"full"</span>,</span>
<span id="cb37-26"><a href="#cb37-26" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="08_stats_models_files/figure-html/cell-26-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can see that there is pretty good overlap between the clusters assigned by the k-means algorithm with the true species groupings. In most cases, you won’t get clusters that perfectly represent your data. For this reason, there are a ton of techniques to help you evaluate and optimize clustering and other classification methods that we won’t get into here. Instead, we will keep it super simple and just look at the proportion of cluster labels for each species. This will give us a rough idea about the quality of the clustering results.</p>
<div id="96e9b657" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> proportions(clusters):</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Count occurrences of each cluster value</span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>    counts <span class="op">=</span> Counter(clusters)</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get total number of items in clusters</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>    total <span class="op">=</span> <span class="bu">len</span>(clusters)</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Calculate proportion for each cluster and sort by cluster id</span></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>    proportions <span class="op">=</span> {cluster: count <span class="op">/</span> total <span class="cf">for</span> cluster, count <span class="kw">in</span> <span class="bu">sorted</span>(counts.items())}</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> proportions</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a><span class="co"># For each species in the iris dataset:</span></span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a><span class="co"># - Assign the cluster labels to a new column 'Cluster'</span></span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a><span class="co"># - Group by 'Species'</span></span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a><span class="co"># - Aggregate the 'Cluster' column using the proportions function</span></span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a><span class="co">#</span></span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a><span class="co"># This shows the distribution of cluster assignments within each species</span></span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a>iris.assign(Cluster<span class="op">=</span>labels).groupby(<span class="st">"Species"</span>).agg({<span class="st">"Cluster"</span>: proportions})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="26">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Cluster</th>
</tr>
<tr class="even">
<th data-quarto-table-cell-role="th">Species</th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">setosa</td>
<td>{1: 1.0}</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">versicolor</td>
<td>{0: 0.06, 2: 0.94}</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">virginica</td>
<td>{0: 0.74, 2: 0.26}</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>That’s pretty good! The k-means clustering pretty well describes the <em>setosa</em> and <em>versicolor</em> species, but the <em>virginica</em> species gets mislabelled about 25% of the time.</p>
<div id="tip-08-clustering-challenges" class="callout callout-style-default callout-tip callout-titled" title="Stop &amp; Think">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="true" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip&nbsp;8.10: Stop &amp; Think
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse show">
<div class="callout-body-container callout-body">
<p>What might explain why k-means clustering performed well for <em>setosa</em> and <em>versicolor</em> but less well for <em>virginica</em>?</p>
</div>
</div>
</div>
</section>
</section>
<section id="summary" class="level2">
<h2 class="anchored" data-anchor-id="summary">Summary</h2>
<p>This chapter covered key statistical and modeling techniques in Python used in life sciences. We looked at methods for comparing groups, including t-tests, ANOVA, and non-parametric alternatives, with a focus on understanding p-values and effect sizes. We also introduced linear regression for exploring relationships between variables, as well as dimensionality reduction and clustering for identifying patterns in data. Throughout, we highlighted the importance of understanding the assumptions behind these methods and interpreting results carefully. Effective data analysis goes beyond running code–it involves choosing the right methods, making sense of the results, and linking them to biological questions. Python offers a range of tools that, when used thoughtfully, support meaningful and clear analysis of experimental data.</p>
</section>
<section id="practice-problems" class="level2">
<h2 class="anchored" data-anchor-id="practice-problems">Practice Problems</h2>
<section id="sec-problem-8.1" class="level3">
<h3 class="anchored" data-anchor-id="sec-problem-8.1">8.1</h3>
<p>Write code to generate two groups of 30 samples each, where group A has a mean of 10 and standard deviation of 2, and group B has a mean of 15 and standard deviation of 2. Then perform an unpaired t-test to compare them.</p>
</section>
<section id="sec-problem-8.2" class="level3">
<h3 class="anchored" data-anchor-id="sec-problem-8.2">8.2</h3>
<p>Add a third group (mean 11, standard deviation of 2) to the dataset from the last problem. Then perform an ANOVA to determine if there are statistically significant differences between them.</p>
</section>
<section id="sec-problem-8.3" class="level3">
<h3 class="anchored" data-anchor-id="sec-problem-8.3">8.3</h3>
<p>Run Tukey’s HSD on the data from <a href="#sec-problem-8.2" class="quarto-xref"><span>Section 8.7.2</span></a>.</p>
</section>
<section id="sec-problem-8.4" class="level3">
<h3 class="anchored" data-anchor-id="sec-problem-8.4">8.4</h3>
<p>Generate a dataset with one response variable and two predictor variables, where only one predictor has a relationship with the response. Fit a multiple regression model and determine which predictor is significant.</p>
</section>
<section id="sec-problem-8.5" class="level3">
<h3 class="anchored" data-anchor-id="sec-problem-8.5">8.5</h3>
<p>Perform k-means clustering with k=2 on the iris dataset. Does this grouping make biological sense? Justify your answer with a visualization.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/mooreryan\.github\.io\/practical_python\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../chapters/07_exploratory_data_analysis.html" class="pagination-link" aria-label="Intro to Exploratory Data Analysis with Python">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Intro to Exploratory Data Analysis with Python</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../chapters/09_io_files_contexts.html" class="pagination-link" aria-label="I/O, Files, &amp; Contexts">
        <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">I/O, Files, &amp; Contexts</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>© 2025 Ryan M. Moore. CC BY 4.0.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://orcid.org/0000-0003-3337-8184">
      <i class="bi bi-person-lines-fill" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://scholar.google.com/citations?user=W08DPpMAAAAJ">
      <i class="bi bi-journals" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/mooreryan">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>
<script async="" src="https://scripts.simpleanalyticscdn.com/latest.js"></script>




</body></html>